src/Computation/Expression.cs

```diff
@@ -2,308 +2,500 @@
+                v => (uint) size switch
+                {
+                    32U => new ConstantSingle((float) (BigInteger) v.AsUnsigned()),
+                    64U => new ConstantDouble((double) (BigInteger) v.AsUnsigned()),
+                    _ => new UnsignedToFloat(size, v)
+                });
         }
 
         public IExpression Write(IExpression offset, IExpression value)
         {
-            return Create(offset, value, (c, o, v) =>
-                c.AsBitVector(_collectionFactory).Write(o.AsUnsigned(), v.AsBitVector(_collectionFactory)));
+            if (Size != offset.Size)
+                throw new InconsistentExpressionSizesException(Size, offset.Size);
+
+            return Ternary(offset, value,
+                (b, o, v) => new Or(
+                    new And(b, new Not(new ShiftLeft(new ZeroExtend(b.Size, ConstantUnsigned.CreateOnes(v.Size)), o))),
+                    new ShiftLeft(new ZeroExtend(b.Size, v), o)),
+                (b, o, v) => b.AsBitVector(_collectionFactory)
+                    .Write(o.AsUnsigned(), v.AsBitVector(_collectionFactory)));
         }
 
         public IExpression Xor(IExpression expression)
         {
-            return Create(expression, (l, r) => l.AsBitwise().Xor(r.AsBitwise()));
+            return Binary(expression,
+                (l, r) => new Xor(l, r),
+                (l, r) => ConstantUnsigned.Create(Size, (BigInteger) l.AsUnsigned() ^ r.AsUnsigned()));
         }
 
         public IExpression ZeroExtend(Bits size)
         {
             return size > Size
-                ? Create(c => c.AsUnsigned().ZeroExtend(size))
+                ? Unary(
+                    v => new ZeroExtend(size, v),
+                    v => v.AsUnsigned().Extend(size))
                 : this;
         }
 
-        private BigInteger AsConstant()
+        private static BigInteger AsConstant(IContextFactory contextFactory, IValue value)
+        {
+            using var handle = contextFactory.Create();
+
+            return value is Float && value.AsFloat(handle.Context).Simplify().IsFPNaN
+                ? value.Size.GetNan(handle.Context)
+                : AsConstant(handle.Context, value);
+        }
+
+        private static BigInteger AsConstant(Context context, IValue value)
         {
-            return _value.AsConstant(_contextFactory);
+            var expr = value.AsBitVector(context).Simplify();
+
+            return expr.IsNumeral
+                ? ((BitVecNum) expr).BigInteger
+                : throw new IrreducibleSymbolicExpressionException();
+        }
+
+        private static IValue GetValue(IPersistentSpace space, IValue value, IValue[] constraints)
+        {
+            using var model = space.GetModel(constraints);
+
+            return ConstantUnsigned.Create(value.Size, model.Evaluate(value));
         }
 
         public static IExpression Create(IContextFactory contextFactory, ICollectionFactory collectionFactory,
```

Comment:
Similarly, I noticed that `constraints` here could also be made non-nullable given an empty enumerable already indicates no values and also in this case it looks like there is currently the possibility of dereferencing a `null` value at `constraints.Select`.